src/zhuyin/syllable.rs

```diff
@@ -0,0 +1,485 @@
+                self.value |= (bopomofo as u16 + 1) << 9;
+            }
+            BopomofoKind::Medial => {
+                if self.step > 1 {
+                    return Err(BuildSyllableError {
+                        msg: "bopomofo is in incorrect order",
+                    });
+                }
+                if self.value & 0b0000_0001_1000_0000 != 0 {
+                    return Err(BuildSyllableError {
+                        msg: "multiple medial bopomofo",
+                    });
+                }
+                self.step = 2;
+                self.value |= (bopomofo as u16 - 20) << 7;
+            }
+            BopomofoKind::Rime => {
+                if self.step > 2 {
+                    return Err(BuildSyllableError {
+                        msg: "bopomofo is in incorrect order",
+                    });
+                }
+                if self.value & 0b0000_0000_0111_1000 != 0 {
+                    return Err(BuildSyllableError {
+                        msg: "multiple rime bopomofo",
+                    });
+                }
+                self.step = 3;
+                self.value |= (bopomofo as u16 - 23) << 3;
+            }
+            BopomofoKind::Tone => {
+                if self.step > 3 {
+                    return Err(BuildSyllableError {
+                        msg: "bopomofo is in incorrect order",
+                    });
+                }
+                if self.value & 0b0000_0000_0000_0111 != 0 {
+                    return Err(BuildSyllableError {
+                        msg: "multiple tone bopomofo",
+                    });
+                }
+                self.step = 4;
+                self.value |= bopomofo as u16 - 36;
+            }
+        };
+        Ok(self)
+    }
+    pub const fn build(self) -> Syllable {
+        Syllable { value: self.value }
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("syllable decode error: {msg}")]
+pub struct DecodeSyllableError {
+    msg: String,
+    source: Box<dyn std::error::Error>,
+}
+
+#[derive(Error, Debug)]
+#[error("syllable build error: {msg}")]
+pub struct BuildSyllableError {
+    msg: &'static str,
+}
+
+#[derive(Error, Debug)]
+#[error("syllable parse error")]
+pub struct ParseSyllableError {
+    source: Box<dyn std::error::Error>,
+}
+
+impl From<ParseBopomofoError> for ParseSyllableError {
+    fn from(value: ParseBopomofoError) -> Self {
+        ParseSyllableError {
+            source: value.into(),
+        }
+    }
+}
+
+impl From<BuildSyllableError> for ParseSyllableError {
+    fn from(value: BuildSyllableError) -> Self {
+        ParseSyllableError {
+            source: value.into(),
+        }
+    }
+}
+
+/// Build a syllable from bopomofos
+///
+/// # Examples
+///
+/// Build a syllable
+/// ```
+/// use chewing::zhuyin::Bopomofo::{K, U, TONE4};
+/// use chewing::syl;
+///
+/// let syl = syl![K, U, TONE4];
+///
+/// assert_eq!("ㄎㄨˋ", syl.to_string());
+/// ```
+#[macro_export]
+macro_rules! syl {
```

Comment:
This macro can panic and it's better to document how it panics since macro doesn't reveal this info in its interface.

Reply:
Good point!